sdks/rust/golem-rust/src/agentic/multimodal.rs

```diff
@@ -167,9 +204,451 @@ impl<T: MultimodalSchema> Multimodal<T> {
+    pub fn custom(value: T) -> CustomModality<T> {
+        CustomModality::Custom(value)
+    }
+}
+
+impl<T: Schema> MultimodalSchema for CustomModality<T> {
+    fn get_multimodal_schema() -> Vec<(String, ElementSchema)> {
+        let mut schema = BasicModality::get_multimodal_schema();
+
+        schema.push((
+            "Custom".to_string(),
+            T::get_type()
+                .get_element_schema()
+                .expect("internal error: unable to get element schema for Custom modality"),
+        ));
+        schema
+    }
+
+    fn get_name(&self) -> String {
+        match self {
+            CustomModality::Basic(basic) => basic.get_name(),
+            CustomModality::Custom(_) => "Custom".to_string(),
+        }
+    }
+
+    fn to_element_value(self) -> Result<(String, ElementValue), String>
+    where
+        Self: Sized,
+    {
+        match self {
+            CustomModality::Basic(basic) => basic.to_element_value(),
+            CustomModality::Custom(custom) => {
+                let elem_value = custom.to_structured_value()?;
+                Ok((
+                    "Custom".to_string(),
+                    elem_value
+                        .get_element_value()
+                        .expect("internal error: unable to get element value for Custom modality"),
+                ))
+            }
+        }
+    }
+
+    fn from_element_value(elem: (String, ElementValue)) -> Result<Self, String>
+    where
+        Self: Sized,
+    {
+        let (name, value) = elem;
+
+        match name.as_str() {
+            "Text" | "Binary" => {
+                let basic = BasicModality::from_element_value((name, value))?;
+                Ok(CustomModality::Basic(basic))
+            }
+            "Custom" => {
+                let schema = T::get_type();
+                let custom = T::from_structured_value(StructuredValue::Default(value), schema)?;
+                Ok(CustomModality::Custom(custom))
+            }
+            _ => Err(format!("Unknown modality name: {}", name)),
+        }
+    }
+
+    fn from_wit_value(wit_value: (String, WitValue)) -> Result<Self, String>
+    where
+        Self: Sized,
+    {
+        let (name, value) = wit_value;
+
+        match name.as_str() {
+            "Text" | "Binary" => {
+                let basic = BasicModality::from_wit_value((name, value))?;
+                Ok(CustomModality::Basic(basic))
+            }
+            "Custom" => {
+                let schema = T::get_type();
+                let custom = T::from_wit_value(value, schema)?;
+                Ok(CustomModality::Custom(custom))
+            }
+            _ => Err(format!("Unknown modality name: {}", name)),
+        }
+    }
+
+    fn to_wit_value(self) -> Result<WitValue, String>
+    where
+        Self: Sized,
+    {
+        match self {
+            CustomModality::Basic(basic) => basic.to_wit_value(),
+            CustomModality::Custom(custom) => custom.to_wit_value(),
+        }
+    }
```

Comment:
The boiler plate of having to define schema can actually be avoided if use type aliases for new variants of advanced multimodal. However, we lose clean API boundaries and user will end up having to see the advanced variants especially when relying on ide to auto complete type. I don't think they ever need to get exposed to advanced variants when they are using any of these simple variants.